### PR TITLE
Add stable public CLI

### PR DESCRIPTION
**Pull request by gonzalobenegas**
PR #94 is merged; this is now the bottom open PR in the modernization stack and should be compared against `main`. Keep it unmerged until reviewed.

## Summary

- replace the legacy per-operation entry points with one typed Cyclopts tree: `gpn ss ...`, `gpn msa ...`, and `gpn star ...`
- expose pinned Transformers `TrainingArguments` directly for inference while enforcing one-row-per-input, input-order, no-drop prediction invariants
- consolidate VEP, logits, and embedding adapters into one `inference.py` per family and compose them with a shared runner/checkpoint implementation
- keep explicit Hugging Face AutoClass registration and remove import-side-effect/custom-loading compatibility paths
- convert the maintained GPN and GPN-Star training profiles to human-readable YAML and keep training entry points intentionally lean
- make resumable inference identity include effective Trainer settings, ordered MSA resources, model/source state, and effective tokenizer serialization
- enforce canonical biallelic SNVs, positive one-based VEP/logits coordinates, reference agreement, boundary padding, even centered MSA windows, and exact odd/even embedding-center sizes
- preserve GPN-MSA as inference-only and keep PhyloGPN/Sorghum inference AutoClass-only

The numeric GPN-Star MSA ordering and exact odd embedding-center size are deliberate correctness fixes covered by focused regression tests.

## Verification

- `UV_CACHE_DIR=/tmp/gpn-uv-cache .venv/bin/pre-commit run --all-files`
- `UV_CACHE_DIR=/tmp/gpn-uv-cache .venv/bin/pytest -q` — 189 passed, 6 intentional skips
- wheel and sdist build, `check-wheel-contents`, and `twine check`
- isolated base-wheel smoke without `accelerate`: `gpn --version` and `gpn --help`
- independent engineering and scientific re-reviews are clean after amendments

No recurring Hugging Face checks and no MSA downloads are added.

### CODE REVIEW

File: docs/cli.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+## GPN-MSA
+
+GPN-MSA is deprecated and supports inference only. Its public commands cover
+variant scoring, masked-nucleotide logits, and embeddings:
+
+```bash
+gpn msa vep INPUT_PATH LOCAL_MSA_PATH WINDOW_SIZE MODEL_PATH OUTPUT_PATH
+gpn msa logits INPUT_PATH LOCAL_MSA_PATH WINDOW_SIZE MODEL_PATH OUTPUT_PATH
+gpn msa embedding INPUT_PATH LOCAL_MSA_PATH WINDOW_SIZE MODEL_PATH OUTPUT_PATH
```
[the remark this answers is not in the document]
Comment (gonzalobenegas) [reply docs/cli.md]: Fixed in 9ea3d50 on this PR. The CLI now forwards center_window_size only when explicitly supplied, preserving EmbeddingInference's default of 100. Added tests for both the omitted and explicit cases; both pass.

File: recipes/gpn_star_training/README.md [hunk 2]
```diff
@@ -63,14 +63,14 @@ Edit the two local alignment paths in `cpu-smoke.json`, then verify the complete
 Hub-dataset/MSA plumbing with one small CPU step:
 
 ```bash
-uv run python -m gpn.star.train recipes/gpn_star_training/cpu-smoke.json
+uv run gpn star train recipes/gpn_star_training/cpu-smoke.json
 ```
 
 For a realistic four-GPU starting point, edit `gpu.json` and run:
```
Comment (gonzalobenegas) [inline recipes/gpn_star_training/README.md]: json is not a very human-friendly format. would yaml be better?

File: docs/cli.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+```bash
+gpn ss vep variants.parquet genome.fa.gz 512 songlab/gpn-brassicales scores.parquet \
+  --is-file --per-device-batch-size 64
+```
+
+The same family group exposes masked-nucleotide logits and sequence embeddings:
+
+```bash
+gpn ss logits POSITIONS_PATH GENOME_PATH WINDOW_SIZE MODEL_PATH OUTPUT_PATH
```
Comment (gonzalobenegas) [inline docs/cli.md]: didn't we discuss a more modern CLI args approach? Is that for later?

File: src/gpn/msa/embedding.py [hunk 4]
```diff
@@ -5,13 +5,13 @@
...
 
-register_auto_classes("msa")
-
 
 class ModelCenterEmbedding(torch.nn.Module):
     def __init__(self, model_path, center_window_size):
         super().__init__()
+        register_auto_classes("msa")
         self.model = AutoModel.from_pretrained(model_path)
+        self.model.eval()
```
Comment (gonzalobenegas) [inline src/gpn/msa/embedding.py]: I'm curious here. do we need autoclasses, vs. just use the explicit class?

File: src/gpn/msa/inference.py [hunk 5]
```diff
@@ -1,100 +1,121 @@
...
+    module_name, class_name = class_mapping[command].split(":", maxsplit=1)
+    return getattr(importlib.import_module(module_name), class_name)
+
+
 def run_inference(
     dataset,
     inference,
     per_device_batch_size=8,
     dataloader_num_workers=0,
+    fp16=None,
```
Comment (gonzalobenegas) [inline src/gpn/msa/inference.py]: I'd like to refactor inference logic. e.g. the transformers prediction loop is pretty general to all, and shouldn't have isolated arguments, but maybe a dataclass of arguments? Which could be like bf16_full_eval, compile, etc...

File: src/gpn/star/inference.py
Comment (gonzalobenegas) [inline src/gpn/star/inference.py:1]: there's so much shared inference logic between models. is now a good time to refactor?

File: recipes/gpn_star_training/README.md (hunk 2 above)
Comment (gonzalobenegas) [reply recipes/gpn_star_training/README.md]: Implemented. Both maintained training recipe families are now YAML-only; the JSON profiles were removed, and parsing, documentation, and tests were updated accordingly.

File: docs/cli.md (hunk 3 above)
Comment (gonzalobenegas) [reply docs/cli.md]: Implemented with Cyclopts and typed Python signatures. Inference commands expose the pinned Transformers TrainingArguments flags directly (for example, --bf16-full-eval), plus a small checkpoint dataclass; there is no --trainer.* nesting. Invalid option combinations now produce a concise CLI error.

File: src/gpn/msa/embedding.py (hunk 4 above)
Comment (gonzalobenegas) [reply src/gpn/msa/embedding.py]: We keep AutoClasses deliberately for standard Hugging Face checkpoint/config dispatch and a uniform user-facing loading API across families. Registration is now explicit via register_auto_classes("msa") immediately before AutoModel/AutoModelForMaskedLM loading—there are no import-side-effect shims or custom model loader.

File: src/gpn/msa/inference.py (hunk 5 above)
Comment (gonzalobenegas) [reply src/gpn/msa/inference.py]: Implemented. gpn.inference now owns the shared Transformers prediction runner, effective TrainingArguments invariants, complete resume-safe runtime fingerprint, atomic output, and checkpoint orchestration. Family code only supplies its model/data adapter and postprocessing.

File: src/gpn/star/inference.py
Comment (gonzalobenegas) [reply src/gpn/star/inference.py:1]: Implemented in this PR. SS, MSA, and Star now compose the same shared inference/checkpoint runner; family-specific VEP, logits, and embedding adapters are consolidated in one inference.py per family. The small internal adapter base classes were also removed in favor of composition and module helpers.

File: README.md [hunk 6]
```diff
@@ -107,30 +124,17 @@ The maintained [GPN training recipe](https://github.com/songlab-cal/gpn/tree/mai
...
+Input requires `chrom`, one-based `pos`, and distinct uppercase canonical SNV
+`ref`/`alt` columns. A mismatch between `ref` and the local genome is an error.
 
 Example command:
 ```bash
-torchrun --nproc_per_node=$(echo $CUDA_VISIBLE_DEVICES | awk -F',' '{print NF}') -m gpn.ss.run_vep \
-    variants.parquet genome.fa.gz 512 your_output_dir results.parquet \
-    --per_device_batch_size 4000 --is_file --dataloader_num_workers 16
+gpn ss vep variants.parquet genome.fa.gz 512 songlab/gpn-brassicales results.parquet \
+    --per-device-eval-batch-size 64 --is-file --dataloader-num-workers 8 \
```
Comment (gonzalobenegas) [inline README.md]: I always felt like the "--is-file" API was weird.

File: README.md [hunk 7]
```diff
@@ -107,30 +124,17 @@ The maintained [GPN training recipe](https://github.com/songlab-cal/gpn/tree/mai
...
-Input file requires `chrom`, `pos`, `ref`, `alt` columns.
+Input requires `chrom`, one-based `pos`, and distinct uppercase canonical SNV
+`ref`/`alt` columns. A mismatch between `ref` and the local genome is an error.
 
 Example command:
 ```bash
-torchrun --nproc_per_node=$(echo $CUDA_VISIBLE_DEVICES | awk -F',' '{print NF}') -m gpn.ss.run_vep \
-    variants.parquet genome.fa.gz 512 your_output_dir results.parquet \
-    --per_device_batch_size 4000 --is_file --dataloader_num_workers 16
+gpn ss vep variants.parquet genome.fa.gz 512 songlab/gpn-brassicales results.parquet \
```
Comment (gonzalobenegas) [inline README.md]: Is there no longer need for torchrun? Also, would it be easier to read if we had something like `--window-size 512` and also each arg on its own line?

File: README.md (hunk 6 above)
Comment (gonzalobenegas) [reply README.md]: Removed the public `--is-file` flag in 4c06e10. Existing local table paths are now detected automatically; recognized table-like paths that do not exist fail with a concise error, while other values are passed to Hugging Face Datasets as dataset identifiers. CLI help and dispatch tests cover the new behavior.

File: README.md (hunk 7 above)
Comment (gonzalobenegas) [reply README.md]: Updated in 4c06e10. Inference inputs are now named flags and the examples put each argument on its own line. `torchrun` remains the supported multi-GPU path; the CLI guide now has a separate multi-GPU inference example (the README quick start stays single-GPU). The shared runner has every rank participate in prediction while rank zero alone commits checkpoints and final output; focused tests and a real 2-process CPU torchrun smoke passed.

File: docs/cli.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+  --window-size 512 \
+  --model-path songlab/gpn-brassicales \
+  --output-path scores.parquet \
+  --per-device-eval-batch-size 64 \
+  --bf16-full-eval
+```
+
+`--per-device-eval-batch-size` is per process, so multiply it by the number of
+processes for the nominal global batch size. Durable checkpoints use the same
+process-count-independent row ranges in single- and multi-GPU runs.
```
Comment (gonzalobenegas) [inline docs/cli.md:206]: could we clarify the logic about number of CPU dataloader workers? Is it per-gpu or total? Also, do we need OMP_NUM_THREADS or not really?

Comment (gonzalobenegas) [reply docs/cli.md:206]: Clarified in f63c263. `--dataloader-num-workers` is per process/rank (normally per GPU under `torchrun`), not a job-wide total, and defaults to zero. The example now sets 4 explicitly: four local ranks therefore create 16 DataLoader worker processes, plus the four rank processes, so the CPU allocation needs to cover all of them. `OMP_NUM_THREADS` is separate and is not required for correctness; when multiple local processes are launched and it is unset, `torchrun` sets it to 1 to avoid oversubscription. The guide recommends leaving that default unless profiling justifies tuning it. Full local validation, PR checks, and an independent pinned-runtime review pass.
